# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd
df = pd.read_csv('../data/supermarket_sales.csv')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [7]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [8]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [9]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [17]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [20]:
df['gross income'].describe(percentiles=[0.01,0.25,0.5,0.75,0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

Значення 75го персентилю = 22.445250
Це означає, що 75% продажів принесли валового прибутку менше або дорівнює 22.44 доларів.

Тобто якщо розташувати всі транзакції за рівнем валового доходу від найменшого до найбільшого і взяти 75% транзакцій із нижчими доходами, то максимальний валовий дохід у цій вибірці буде 22.445250.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [24]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [25]:
df.City.value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [30]:
yangon_mean = df.query('City == "Yangon"')['gross income'].mean()
mandalay_mean = df.query('City == "Mandalay"')['gross income'].mean()

display(yangon_mean, mandalay_mean)


np.float64(14.874001470588235)

np.float64(15.232024096385542)

In [31]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

Гіпотеза 1: Цінова політика на одні й ті самі товарні позиції різна
При однаковій собівартості товарів, у Mandalay ціни на товари вищі, ніж у Yangon, що призводить до більшого середнього валового доходу.

Як перевірити:
Проаналізувати середню вартість проданих товарів (колонка Unit price).

Додатково можна також дізнатися про локальні ціни на товари для аналогічних товарних груп, податкове навантаження, купівельну спроможність населення, аби зрозуміти чи має сенс для міста Yangon також підвищити ціну. 

In [32]:
yangon_mean_price = df.query('City == "Yangon"')['Unit price'].mean()
mandalay_mean_price = df.query('City == "Mandalay"')['Unit price'].mean()

display(yangon_mean_price, mandalay_mean_price)

np.float64(54.78085294117646)

np.float64(55.65927710843374)

Гіпотеза 2: Різна кількість одиниць товару в одному чеку
При практично однаковій загальній кільксоті транзакцій, може мати вплив на валовий дохід кількість товарів в одному чеку.

Як перевірити:
Проаналізувати середню кількість одиниць товару в чеку (Quantity) для обох міст, щоб зрозуміти, чи впливають продажі більшої кількості товарів на дохід.и

In [33]:
yangon_mean_quantity = df.query('City == "Yangon"')['Quantity'].mean()
mandalay_mean_quantity = df.query('City == "Mandalay"')['Quantity'].mean()

display(yangon_mean_quantity, mandalay_mean_quantity)

np.float64(5.4676470588235295)

np.float64(5.481927710843373)

Гіпотеза 3: Різниця в категоріях товарів, що продаються
У Mandalay частіше купують високомаржинальні товари, тоді як у Yangon переважають продажі товарів із низькою чи середьною маржею.

Як перевірити:
Аналізувати продажі за категоріями товарів (Product line) і порівняти середній дохід від кожної категорії в обох містах.

Додатково можна отримати інформацію про маркетингові стратегії, спеціальні пропозиції або акції, які проводились у кожному місті, аби прийняти рішення щодо просування товарів з вищою маржею в Yangon, використовуючи цільові акції, маркетингові кампанії або рекомендації.

In [44]:
yangon_category_sales = df.query('City == "Yangon"')['Product line'].value_counts()
mandalay_category_sales = df.query('City == "Mandalay"')['Product line'].value_counts()
yangon_category_income = df.query('City == "Yangon"').groupby(['Product line'])['gross income'].mean()
mandalay_category_income = df.query('City == "Mandalay"').groupby(['Product line'])['gross income'].mean()
comparison = pd.DataFrame({"Yangon Sales": yangon_category_sales, "Mandalay Sales": mandalay_category_sales, "Yangon Income": yangon_category_income, "Mandalay Income": mandalay_category_income})
comparison

Yangon Sales  Mandalay Sales  Yangon Income  \
Product line                                                          
Electronic accessories            60              55      14.537392   
Fashion accessories               51              62      15.249775   
Food and beverages                58              50      14.091216   
Health and beauty                 47              53      12.763681   
Home and lifestyle                65              50      16.422854   
Sports and travel                 59              62      15.635754   

                        Mandalay Income  
Product line                             
Electronic accessories        14.763155  
Fashion accessories           12.606234  
Food and beverages            14.490370  
Health and beauty             17.952075  
Home and lifestyle            16.713490  
Sports and travel             15.351919

Середній валовий дохід в усіх категоріях, окрім Fashion acessories & Health and beauty в обох містах практично однаковий, але кількість продажів в місті Yangon більша, що свідчить про те, що товари продаються за нижчими цінами. Дохід від продажу Fashion acessories вищий в місті Yangon в порівнянні з Mandalay при меншій кількості проданих одиниць, тобто ця категорія товарів являється більш маржинальною.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [52]:
gross_margin_unique = df['gross margin percentage'].nunique()

if gross_margin_unique == 1:
    conclusion = "У всіх інвойсах однакова маржа"
else:
    conclusion = "Маржа відрізняється для різних інвойсів"

print(gross_margin_unique, conclusion, sep=': ')

1: У всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [67]:
filtered_cities = df.query('City == "Yangon" | City == "Mandalay"')

normalized_counts = (
    filtered_cities.groupby(['City', 'Product line']).size() /
    filtered_cities.groupby(['City']).size()
).reset_index(name='Percentage')

normalized_counts

City            Product line  Percentage
0   Mandalay  Electronic accessories    0.165663
1   Mandalay     Fashion accessories    0.186747
2   Mandalay      Food and beverages    0.150602
3   Mandalay       Health and beauty    0.159639
4   Mandalay      Home and lifestyle    0.150602
5   Mandalay       Sports and travel    0.186747
6     Yangon  Electronic accessories    0.176471
7     Yangon     Fashion accessories    0.150000
8     Yangon      Food and beverages    0.170588
9     Yangon       Health and beauty    0.138235
10    Yangon      Home and lifestyle    0.191176
11    Yangon       Sports and travel    0.173529

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)